kernel/crates/kernel_physical_memory/src/lib.rs

```diff
@@ -22,106 +25,172 @@ impl FrameState {
-            None
+    fn allocate_frames_impl<S: PageSize>(
+        &mut self,
+        n: usize,
+    ) -> Option<PhysFrameRangeInclusive<S>> {
+        let small_frames_per_frame = (S::SIZE / Size4KiB::SIZE) as usize;
+        let small_frame_count = n * small_frames_per_frame;
+
+        let ff = self.first_free()?;
+
+        // TODO: Support searching across region boundaries for better memory utilization
+        // Search for contiguous free frames within regions
+        for region_idx in ff.region_idx..self.regions.len() {
+            let search_start = if region_idx == ff.region_idx {
+                ff.frame_idx
+            } else {
+                0
+            };
+
+            let region = &self.regions[region_idx];
+            if search_start >= region.len() {
+                continue;
+            }
+
+            // Align search_start up to the required page size alignment
+            let aligned_search_start = {
+                let offset = search_start % small_frames_per_frame;
+                if offset == 0 {
+                    search_start
+                } else {
+                    search_start + (small_frames_per_frame - offset)
+                }
+            };
+
+            // Search for contiguous free frames
+            let mut current_start = aligned_search_start;
+            while current_start + small_frame_count <= region.len() {
+                // Check if we have enough contiguous free frames
+                let all_free = region.frames()[current_start..current_start + small_frame_count]
+                    .iter()
+                    .all(|&state| state == FrameState::Free);
+
+                if all_free {
+                    let frame_start_idx = current_start;
+                    let frame_end_idx = current_start + small_frame_count - 1;
+
+                    // Get the physical addresses before mutating
+                    let start_addr = self.regions[region_idx].frame_address(frame_start_idx)?;
+                    let end_addr_idx =
+                        frame_end_idx / small_frames_per_frame * small_frames_per_frame;
+                    let end_addr = self.regions[region_idx].frame_address(end_addr_idx)?;
+
+                    // Mark frames as allocated
+                    self.regions[region_idx].frames_mut()[frame_start_idx..=frame_end_idx]
+                        .fill(FrameState::Allocated);
+
+                    // Update first_free pointers
+                    if region_idx == ff.region_idx && frame_start_idx <= ff.frame_idx {
+                        self.update_first_free(region_idx, frame_end_idx + 1);
+                    }
+
+                    // Convert to physical frames
+                    return Some(PhysFrameRangeInclusive {
+                        start: PhysFrame::from_start_address(PhysAddr::new(start_addr)).ok()?,
+                        end: PhysFrame::from_start_address(PhysAddr::new(end_addr)).ok()?,
+                    });
```

Comment:
Instead of `.ok()?`, panic with `expect`, since this is developer error, and because we use an incorrect address, doesn't mean we're out of memory.

Also, please add a test, where we allocate a for example 2MiB frame but the first free 2MiB "slot" is not 2MiB aligned. Currently, this causes an incorrect out of memory error, because the start address is not aligned.